ansible_base/authentication/social_auth.py

```diff
@@ -189,12 +192,57 @@ def validate(self, serializer, data):
         return data
 
 
+def extract_id_token_claims(backend, response):
+    try:
+        id_token = response['id_token']
+        client_id, _ = backend.get_key_and_secret()
+
+        key = backend.find_valid_key(id_token)
+
+        if not key:
+            raise AuthTokenError(backend, "Signature verification failed")
+
+        rsakey = jwt.PyJWK(key)
+        id_token_claims = jwt.decode(
+            id_token,
+            rsakey.key,
+            algorithms=backend.setting("JWT_ALGORITHMS", backend.JWT_ALGORITHMS),
+            audience=client_id,
+            issuer=backend.id_token_issuer(),
+            options=backend.setting("JWT_DECODE_OPTIONS", backend.JWT_DECODE_OPTIONS),
+        )
+    except PyJWTError as e:
+        logger.error(f"Unable to decode id_token response JWT: {e}")
+        id_token_claims = {}
+    except AuthTokenError as e:
+        logger.error(f"Unable to decode id_token response JWT: {e}")
+        id_token_claims = {}
+
+    return id_token_claims
+
+
 def create_user_claims_pipeline(*args, backend, response, **kwargs):
     from ansible_base.authentication.utils.claims import update_user_claims
 
     groups_claim = backend.groups_claim if backend.groups_claim is not None else "Group"
 
-    extra_groups = response[groups_claim] if groups_claim in response else []
+    if groups_claim in response:
+        # Attempt to extact groups claim from data returned by UserInfo endpoint
+        extra_groups = response[groups_claim]
+        logger.debug(f"extra_groups extracted from UserInfo: {extra_groups}")
+    else:
+        # If groups_claim isn't in UserInfo attempt to extract it from id_token
+        if 'id_token' in response:
+            id_token_claims = extract_id_token_claims(backend, response)
+        else:
+            id_token_claims = {}
+
+        if groups_claim in id_token_claims:
+            extra_groups = id_token_claims[groups_claim]
+            logger.debug(f"extra_groups extracted from id_token: {extra_groups}")
+        else:
+            extra_groups = []
```

Comment:
What are your thoughts on simplifying this block a bit:
```suggestion
        extra_groups = id_token_claims.get(groups_claim, [])
        logger.debug(f"extra_groups extracted from id_token: {extra_groups}")
```

This would also give us the benefit of always logging from this function so we would know that we got into here even if we ended up with no groups.

Reply:
seems reasonable

Reply:
fixed